test/teststl.cpp

```diff
@@ -3844,6 +3845,81 @@ class TestStl : public TestFixture {
               true);
         ASSERT_EQUALS("[test.cpp:4]: (style) Consider using std::accumulate algorithm instead of a raw loop.\n", errout.str());
     }
+
+    void findInsert()
+    {
+        check("void f1(std::set<unsigned>& s, unsigned x) {\n"
+              "    if (s.find(x) == s.end()) {\n"
+              "        s.insert(x);\n"
+              "    }\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("[test.cpp:3]: (performance) Searching before insertion is not necessary.\n", errout.str());
+
+        check("void f2(std::map<unsigned, unsigned>& m, unsigned x) {\n"
+              "    if (m.find(x) == m.end()) {\n"
+              "        m.emplace(x, 1);\n"
+              "    } else {\n"
+              "        m[x] = 1;\n"
+              "    }\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("[test.cpp:3]: (performance) Searching before insertion is not necessary.\n", errout.str());
+
+        check("void f3(std::map<unsigned, unsigned>& m, unsigned x) {\n"
+              "    if (m.count(x) == 0) {\n"
+              "        m.emplace(x, 1);\n"
+              "    }\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("[test.cpp:3]: (performance) Searching before insertion is not necessary.\n", errout.str());
```

Comment:
I fear that the logic is changed if the search is removed here. Imagine that `m[x]` is `1000`. Without the search the value is changed.

Reply:
No, its not. `emplace` or `insert` does not insert a new value if it is already found.

Reply:
I see.